### PR TITLE
WT-10469 Fix no timestamp tombstones not removing history store entries

### PR DESCRIPTION
**Pull request by Sean04**
The scenario that this fixes is outlined in the ticket. This change also includes some refactoring to make calls to `__wt_rec_hs_clear_on_tombstone` more consistent. More scenarios were also added to test_hs11 that reproduce the original bug by forcing reconciliation to occur. The scenarios with modify updates and long-running transactions also uncovered the bug which WT-10512 fixes.

### CODE REVIEW

File: src/reconcile/rec_row.c [hunk 1]
```diff
@@ -866,23 +880,26 @@ __wt_rec_row_leaf(
...
-                if (upd_select.no_ts_tombstone && r->hs_clear_on_tombstone) {
-                    WT_ERR(__wt_row_leaf_key(session, page, rip, tmpkey, true));
-                    WT_ERR(__wt_rec_hs_clear_on_tombstone(session, r, WT_RECNO_OOB, tmpkey, false));
-                }
-
                 /* Not creating a key so we can't use last-key as a prefix for a subsequent key. */
                 lastkey->size = 0;
-
-                /* Proceed with appended key/value pairs. */
-                goto leaf_insert;
```
Comment (quchenhao) [inline src/reconcile/rec_row.c:882]: Is it correct to not skip the code anymore?

File: src/reconcile/rec_row.c [hunk 2]
```diff
@@ -866,23 +880,26 @@ __wt_rec_row_leaf(
...
+             * versions that are greater in the history store for this key.
+             */
+            if (upd_select.no_ts_tombstone && r->hs_clear_on_tombstone) {
+                WT_ERR(__wt_row_leaf_key(session, page, rip, tmpkey, true));
+                WT_ERR(__wt_rec_hs_clear_on_tombstone(session, r, WT_RECNO_OOB, tmpkey,
+                  upd->type == WT_UPDATE_TOMBSTONE ? false : true));
+            }
+
+            /* Proceed with appended key/value pairs. */
+            if (upd->type == WT_UPDATE_TOMBSTONE)
```
Comment (quchenhao) [inline src/reconcile/rec_row.c:901]: I don't really like the refactor here. Would be better just duplicate the code in two places? @hbokhari

File: src/reconcile/rec_row.c [hunk 3]
```diff
@@ -557,25 +562,40 @@ __rec_row_leaf_insert(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins)
...
+         * that are greater in the history store for this key.
+         */
+        if (upd_select.no_ts_tombstone && r->hs_clear_on_tombstone) {
+            tmpkey->data = WT_INSERT_KEY(ins);
+            tmpkey->size = WT_INSERT_KEY_SIZE(ins);
+            WT_ERR(__wt_rec_hs_clear_on_tombstone(
+              session, r, WT_RECNO_OOB, tmpkey, upd->type == WT_UPDATE_TOMBSTONE ? false : true));
         }
+
+        if (upd->type == WT_UPDATE_TOMBSTONE)
```
Comment (quchenhao) [inline src/reconcile/rec_row.c:594]: Same here.

File: src/reconcile/rec_row.c (hunk 2 above)
Comment (kommiharibabu) [reply src/reconcile/rec_row.c:901]: We need to copy the same code into three places. Modify, standard update and tombstone. Instead bringing it down will have only one copy.

File: test/suite/test_hs11.py [hunk 4]
```diff
@@ -44,8 +44,18 @@ class test_hs11(wttest.WiredTigerTestCase):
         ('deletion', dict(update_type='deletion')),
         ('update', dict(update_type='update'))
     ]
-    scenarios = make_scenarios(format_values, update_type_values)
+    long_running_txn_values = [
+       ('long-running', dict(long_run_txn=True)),
```
Comment (kommiharibabu) [inline test/suite/test_hs11.py:48]: This test doesn't pass until WT-10512 is merged. we need to decide the merge order to enable it.

Comment (Sean04) [reply test/suite/test_hs11.py:48]: If we merge that first, this test should pass.

File: test/suite/test_hs11.py [hunk 5]
```diff
@@ -115,7 +159,97 @@ def test_non_ts_updates_clears_hs(self):
...
+                        if ts == 5 and self.modify and self.value_format != '8t':
+                            self.assertEqual(cursor[self.create_key(i)], mod_value)
+                        else:
+                            self.assertEqual(cursor[self.create_key(i)], value1)
+
+        if self.update_type == 'deletion':
+            hs_truncate = self.get_stat(stat.conn.cache_hs_key_truncate_onpage_removal)
+            self.assertGreater(hs_truncate, 0)
+
+    def test_non_ts_updates_clears_hs_small_rows(self):
```
Comment (Sean04) [inline test/suite/test_hs11.py:171]: @kommiharibabu I was thinking we could change this to be a scenario of the above test. Thoughts?

Comment (kommiharibabu) [reply test/suite/test_hs11.py:171]: Nice idea. Yes, by controlling the number of rows based on the scenario. It will simplify the test.

**Review by kommiharibabu — APPROVED** [review]
LGTM. Thanks.

**Review by quchenhao — APPROVED** [review]
lgtm

**Comment by Sean04** [comment]
evergreen merge

**Pull request merged**